paddle/framework/recurrent_network_op.h

```diff
@@ -0,0 +1,141 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#pragma once
+
+#include "paddle/framework/enforce.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/variable.h"
+
+namespace paddle {
+namespace framework {
+
+// fake interfaces that has not be implemented by other modules.
+struct OpRunContext {
+  Scope* scope;
+};
+
+// TODO replace this with Net's proto.
+struct NetDesc {
+  std::string name;
+}
+
+class OperatorBase {
+ public:
+  virtual ~OperatorBase() {}
+  virtual void Run(OpRunContext* context) const = 0;
+  virtual void InferShape(const Scope* scope) const = 0;
+
+ protected:
+  std::vector<std::string> inputs_;
+  std::vector<std::string> outputs_;
+}
+
+class RecurrentGroupForwardOp {
+ public:
+  RecurrentGroupForwardOp(NetDesc& net_desc)
+      : name_(net_desc.name),
+        net_name_(net_desc.name + "__net__"),
+        step_scopes_name_(net_desc.name + "__step_scopes_") {}
+
+  virtual void InferShape(const Scope* scope) = 0;
+  /*
+   * Forward run the RNN.
+   *
+   * NOTE the context's scope is not given until `Run` called, so step scopes'
+   * father should be set/updated in this method.
+   */
+  virtual void Run(OpRunContext* contex) const {
```

Comment:
Yes, will move to .cpp later.

We are working on a simple implementation to verify the whole process and will give a version soon.